### ISSUE TITLE
--link hangs when source is unavailable.

### ISSUE THREAD

**Issue by moxiegirl** [docker/docker#21834]
**Output of `docker version`:**

```
docker version
Client:
 Version:      1.10.3
 API version:  1.22
 Go version:   go1.5.3
 Git commit:   20f81dd
 Built:        Thu Mar 10 21:49:11 2016
 OS/Arch:      darwin/amd64

Server:
 Version:      1.10.3
 API version:  1.22
 Go version:   go1.5.3
 Git commit:   20f81dd
 Built:        Thu Mar 10 21:49:11 2016
 OS/Arch:      linux/amd64
```

**Output of `docker info`:**

```
unavailable
```

**Additional environment details (AWS, VirtualBox, physical, etc.):**

I was running with docker-machine so the host was created on machine. But Machine doesn't appear to be the problem.

**Steps to reproduce the issue:**
1. Run a container and specify a `--link` where the source does not exist.  For example:

```
docker run -ti -d \
→       -p 80:80 \
→       --label interlock.ext.name=nginx \
→       --link=nginx:interlock \
→       -v nginx:/etc/conf \
→       --name nginx \
→       nginx nginx -g "daemon off;" -c /etc/conf/nginx.conf
de88214aefb17c0343a58f763545ca3c8b72878fa764b76383723ccccee26e35
```

**Describe the results you received:**

The `docker run` outputs a shell for the container I'm creating but then hangs. 

**Describe the results you expected:**

If the link source doesn't exist, the system should fail the run with an error. As it stands now, any docker command against that container just hangs.  I can't delete it even tho I have the SHA.

Repro'd on another machine.

**Additional information you deem important (e.g. issue happens only occasionally):**

**Comment by cpuguy83** [comment]
I cannot repro, at least not on 1.11.

**Comment by moxiegirl** [comment]
@cpuguy83  I'll try to repro there.

**Comment by moxiegirl** [comment]
@cpuguy83 are you creating through `docker-machine` or just `docker`?  I was able to repro this under 1.11 when running `docker-machine`

```
$ docker run  --link=nginx:interlock --name nginx nginx
Unable to find image 'nginx:latest' locally
latest: Pulling from library/nginx
efd26ecc9548: Pull complete 
a3ed95caeb02: Pull complete 
83f52fbfa5f8: Pull complete 
fa664caa1402: Pull complete 
Digest: sha256:12127e07a75bda1022fbd4ea231f5527a1899aad4679e3940482db3b57383b1d
Status: Downloaded newer image for nginx:latest
```

On Ubutu trusty

```
$ docker run --name nginx -d nginx
docker: Cannot connect to the Docker daemon. Is the docker daemon running on this host?.
See 'docker run --help'.
ubuntu@ubuntu1404:~$ sudo docker run  --link=nginx:interlock --name nginx nginx
Unable to find image 'nginx:latest' locally
latest: Pulling from library/nginx
efd26ecc9548: Pull complete 
a3ed95caeb02: Pull complete 
83f52fbfa5f8: Pull complete 
fa664caa1402: Pull complete 
Digest: sha256:12127e07a75bda1022fbd4ea231f5527a1899aad4679e3940482db3b57383b1d
Status: Downloaded newer image for nginx:latest
```

On both, a `docker ps` just hangs after running the nginx as shown.

```
ubuntu@ubuntu1404:~$ sudo docker images
REPOSITORY          TAG                 IMAGE ID            CREATED             SIZE
nginx               latest              eb4a127a1188        12 days ago         182.6 MB
hello-world         latest              690ed74de00f        6 months ago        960 B
ubuntu@ubuntu1404:~$ sudo docker ps
```

**Comment by moxiegirl** [comment]
btw @nathanleclaire took a look at this with me last Friday.  He's thinking Engine bug as well.

**Comment by cpuguy83** [comment]
I think this is fixed on master.

**Comment by cpuguy83** [comment]
Also, a dup of #20271

**Comment by AkihiroSuda** [comment]
the issue seems fixed in upstream (ccd0d83344da3b4c67708465f11aa542601b6407), not sure which commit fixed that..

```
$ docker run  --link=nginx:interlock --name nginx nginx
Unable to find image 'nginx:latest' locally
latest: Pulling from library/nginx
51f5c6a04d83: Already exists 
a3ed95caeb02: Pull complete 
51d229e136d0: Pull complete 
bcd41daec8cc: Pull complete 
Digest: sha256:0fe6413f3e30fcc5920bc8fa769280975b10b1c26721de956e1428b9e2f29d04
Status: Downloaded newer image for nginx:latest
docker: Error response from daemon: Could not get container for nginx.
See 'docker run --help'.
```

Is this closable?

**Comment by thaJeztah** [comment]
confirmed, I'm not able to reproduce this on 1.12
